### DISCUSSION TITLE
Clarify caching hits for dependsOn

### DISCUSSION THREAD

**Discussion by RMcNeely** [vercel/turborepo#2962] [Help] [answered]
Our team is working on implementing Turbo repo and so far we've had a lot of success with improving our linting and test jobs but not so much for builds. I want to clarify that I am setting up the relationships between my packages correctly and that I should (probably) not expect to see much of a difference in build times due to having caches being hit.

Our app is an ejected CRA that we have been converting to a traditional monorepo where each team gets their own package. At this point we have three packages. For convenience let's call them `Foo`, `Bar`, and `Legacy`. In our scenario Foo is the entry point of our build process (Webpack) where we do the final combinations and handle different file formats and final transpiling. It has a dependency on Bar which is our first ejected package, this is specified in the package.json file for Foo. Legacy is one folder with all the rest of our code that is imported by Foo. For the now it can be safely ignored.

My turbo.json file has a build pipeline that looks like this:
```json
  "build": {
    "dependsOn": ["^build"]
  }
```

Originally I expected that this relationship would mean that if I made a change only to _package/Bar_ that I would only invalidate the cache for `Bar` and would only ever rebuild that. In practice however I notice that when I make changes to _packages/Bar_ I am also missing the cache for packages/Foo and end up building both apps at no time savings. When I do make change only to `Foo` or `Legacy` then I do not rebuild `Bar`.

I assume that this is because my Foo package depends on Bar and the change to Bar is invalidating the cache of Foo forcing them to both rebuild. Is that correct? Is there a way to configure Turbo to not always rebuild Foo when I change Bar or is that always  going to be the case because Bar is listed as a dependency?

**Comment by anthonyshew** [comment, marked as answer]
It sounds like the understanding described here is correct. Restating in my own words:
- Packages are inputs to apps.
- If your package changes, you want your app to rebuild (so that it reflects those changes).
- The `^` character in particular notifies the Task Graph of this relationship. It specifies dependence in the topological tree.
